webhookClient/basicRegisterer.go

```diff
@@ -129,39 +129,65 @@ func NewBasicRegisterer(acquirer Acquirer, secret SecretGetter, config BasicConf
 func (b *BasicRegisterer) Register() error {
 	secret, err := b.secretGetter.GetSecret()
 	if err != nil {
-		return emperror.Wrap(err, "Failed to get secret")
+		return errWithReason{
+			err:    emperror.Wrap(err, "Failed to get secret"),
+			reason: GetSecretFail,
+		}
 	}
 
 	b.requestTemplate.Config.Secret = secret
 
 	marshaledBody, errMarshal := json.Marshal(&b.requestTemplate)
 	if errMarshal != nil {
-		return emperror.WrapWith(errMarshal, "failed to marshal")
+		return errWithReason{
+			err:    emperror.WrapWith(errMarshal, "failed to marshal"),
+			reason: MarshalRequestFail,
+		}
 	}
 
-	satToken, err := b.acquirer.Acquire()
+	jwtToken, err := b.acquirer.Acquire()
 	if err != nil {
-		return err
+		return errWithReason{
+			err:    err,
+			reason: AcquireJWTFail,
+		}
 	}
 
-	req, _ := http.NewRequest("POST", b.registrationURL, bytes.NewBuffer(marshaledBody))
-	if satToken != "" {
-		req.Header.Set("Authorization", satToken)
+	req, err := http.NewRequest("POST", b.registrationURL, bytes.NewBuffer(marshaledBody))
+	if err != nil {
+		return errWithReason{
+			err:    err,
+			reason: CreateRequestFail,
+		}
+	}
+
+	if jwtToken != "" {
+		req.Header.Set("Authorization", jwtToken)
 	}
 
 	resp, err := b.client.Do(req)
 	if err != nil {
-		return emperror.WrapWith(err, "failed to make http request")
+		return errWithReason{
+			err:    emperror.WrapWith(err, "failed to make http request"),
```

Comment:
Are we getting anything with the `emperror` package?  The `Wrap...` functions explicitly document that they should not be used when passing errors between goroutines, which seems to be an especially dangerous constraint.  Future code might do exactly that, and the maintenance programmer might not be aware of that constraint.

It might be simpler just to pass the original error and not worry about annotating it.

Reply:
I usually wrap errors so it's easy to trace through the code what returned an error that was logged; I'm open to any better ways to do that though! We don't have to use emperror to do that: the easiest and fastest way I can think of is just make a struct that chains additions to an error aka our own error wrapper.

Reply:
Wrapping errors is totally fine.  If you need to preserve the original error, say if you want callers to do `errors.Unwrap`, then you need to wrap things.  But you don't need `emperror` for that.  There are lots of libraries that do the same thing.  Or you could roll your own.

My question was more driven by the fact that `emperror` does things in the wrapper that have side effects that are likely very undesireable.